onmt/translate/greedy_search.py

```diff
@@ -38,17 +87,10 @@ def sample_with_temperature(logits, sampling_temp, keep_topk):
             topk_scores /= sampling_temp
     else:
         logits = torch.div(logits, sampling_temp)
-
+        if keep_topp > 0:
+            logits = sample_topp(logits, keep_topp, always_sample_eos, eos)
         if keep_topk > 0:
```

Comment:
I don't find it very clear how topp and topk intertwine. Does the original paper intend for both to be used together, and in that order?

Reply:
The order won't change the result, it's like a logical "and".
The original paper does not combine the two methods.
Huggingface allows both methods to be combined. With fairseq it is one or the other.

For example if after topp filtering there is more than k tokens selected, topk will limit the amount of tokens to k. It's a second filtering step.
You can always use only one of the two, it is not necessary to use both at the same time